### PR TITLE
speed-up dataloader iteration in case the dataset is a TensorDataset …

### PR DESCRIPTION
**Pull request by richardalligier**
The iteration of a Dataloader object can be speed-up when the dataset is a TensorDataset object and default_collate is used. In such a case, I have replaced `collate_fn([dataset[i] for i in indices])` by `[tensor.index_select(0, torch_indices) for tensor in dataset.tensors]`.

I used the code below to test it
```
N=10000
xs = torch.randn(N,5)
ys = torch.randn(N)
ds = torch.utils.data.TensorDataset(xs,ys)
train_loader = DataLoader(ds, batch_size=256, shuffle=True, num_workers=NUM_WORKER)
start = time.time()
for _ in range(10):
    for _ in train_loader:
        pass
print(time.time() - start)
```
On my old 2-cores laptop (without GPU), I have compared before and after my modification:

| NUM_WORKER        | Before [s]           | After [s]  |
| ------------- |:-------------:| -----:|
| 0      | 1.39 | 0.09 |
| 1      | 6.01      |   4.07 |
| 2 | 4.71      |    3.49 |

### CODE REVIEW

**Review by fmassa — CHANGES_REQUESTED** [review]

**Comment by fmassa** [comment]
Actually, for your particular case, can't you instead write a new batch sampler that returns a tuple with a single list inside, and the list contains the indices to pass to tensordataset ? I think we don't need to modify anything in the current codebase, nor expose a new functionality

**Comment by richardalligier** [comment]
I agree, this is very specific to a particular kind of datasets however this type of datasets are pretty common for regression problems.

The Dataloader class is very convenient. I only had performance problem with it. Thus, I fixed the performance issues without changing the API nor the behavior (at least I hope so). I could do it with a batch sampler but it felt like I would be writing my own Dataloader.

Having a `batch_collate_fn` parameter would solve my issue and might also be useful for others who have data structure for which 1 query of n elements is faster than n queries of 1 element. Though, I will understand if you don't want to change the API.

**Comment by fmassa** [comment]
I had thought about it in the past, and I think that having a Batch sampler that returns a Batch of indices at once might be the best solution here. It doesn't involve modifying the DataLoader, and also works for arbitrary datasets that support batch indexing.
We could maybe add such option to the default batch sampler?

**Comment by fmassa** [comment]
Let me know if you are unsure on how to proceed and I can give you some more pointers

**Comment by richardalligier** [comment]
Thanks for your help, I think I have understood, is it something like that ?
```
class MyTensorDataset(TensorDataset):
    def __getitem__(self,index):
        if isinstance(index, int):
            return super(MyTensorDataset,self).__getitem__(index)
        else:
            batch_index=torch.LongTensor(index)
            return tuple(tensor.index_select(0, batch_index) for tensor in self.tensors)

def removefirstdim(iterator):
    for tensors in iterator:
        yield tuple(tensor[0] for tensor in tensors)

N=10000
xs = torch.randn(N,5)
ys = torch.randn(N)
ds = MyTensorDataset(xs,ys)
batch_sampler = BatchSampler(sampler=RandomSampler(range(N)), batch_size=256, drop_last=False)
train_loader = removefirstdim(DataLoader(ds,sampler=batch_sampler, num_workers=NUM))
start=time.time()
for _ in range(10):
    for x,y in train_loader:
        pass
print(time.time()-start)
```
I used the `removefirstdim` function to remove the dimension added by the DataLoader. It might not be necessary depending on what is done inside the training loop.

File: torch/utils/data/dataloader.py [hunk 1]
```diff
@@ -189,6 +189,14 @@ def default_collate(batch):
     raise TypeError((error_msg.format(type(batch[0]))))
 
 
+def select_indices_and_collate(collate_fn, dataset, indices):
+    if collate_fn == default_collate and isinstance(dataset, torch.utils.data.TensorDataset):
+        torch_indices = torch.LongTensor(indices)
+        return [tensor.index_select(0, torch_indices) for tensor in dataset.tensors]
```
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/utils/data/dataloader.py]: I'd rather not merge this PR. It complicates the data loader code, for a use case that doesn't really benefit from the abstraction. You can always just use a batch sampler in your loop, and index the tensor manually if you really need this kind of perf.

**Review by apaszke — CHANGES_REQUESTED** [review]

**Comment by richardalligier** [comment]
I agree, there is actually no need to modify the DataLoader. I didn't thought about using the BatchSampler. I can use it with the DataLoader like I did in the code above or even simpler just build a custom iterator using the BatchSampler.

Thanks for your help and time @fmassa and @apaszke .

**Pull request closed without merging**